google/resource_kms_key_ring_test.go

```diff
@@ -0,0 +1,78 @@
+package google
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"log"
+)
+
+func TestAccGoogleKmsKeyRing_basic(t *testing.T) {
+	name := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testGoogleKmsKeyRing_recreate(name),
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testGoogleKmsKeyRing_basic(name),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGoogleKmsKeyRingExists("google_kms_key_ring.key_ring"),
+				),
+			},
+		},
+	})
+}
+
+func testGoogleKmsKeyRing_basic(name string) string {
+	return fmt.Sprintf(`
+	resource "google_kms_key_ring" "key_ring" {
+		name = "%s"
+		location = "us-central1"
+	}
+	`, name)
+}
+
+func testAccCheckGoogleKmsKeyRingExists(resourceName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		config := testAccProvider.Meta().(*Config)
+
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Resource not found: %s", resourceName)
+		}
+
+		name := rs.Primary.Attributes["name"]
+		location := rs.Primary.Attributes["location"]
+
+		parent := createKmsResourceParentString(config.Project, location)
+		keyRingName := createKmsResourceKeyRingName(config.Project, location, name)
+
+		listKeyRingsResponse, err := config.clientKms.Projects.Locations.KeyRings.List(parent).Do()
+		if err != nil {
+			return fmt.Errorf("Error listing KeyRings: %s", err)
+		}
+
+		for _, keyRing := range listKeyRingsResponse.KeyRings {
+			log.Printf("[DEBUG] Found KeyRing: %s", keyRing.Name)
+
+			if keyRing.Name == keyRingName {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("KeyRing not found: %s", keyRingName)
+	}
+}
+
+// TODO
+// KMS KeyRings cannot be deleted. This will test if the resource can be added back to state after being removed
+func testGoogleKmsKeyRing_recreate(name string) resource.TestCheckFunc {
```

Comment:
If the resource already exists server-side, we can add import functionality instead of trying to get it to essentially import on create.

Reply:
What is the expected behavior of attempting to create a KeyRing that already exists?  Should it fail because the resource already exists on the server?  Or should it handle it similar to how an import would?

I think most resources would fail here because you'd expect to have to use `terraform import` for resources not managed by terraform.  This seems like a special situation though, because for all other resources, you can do `terraform destroy` and `terraform apply` to re-create your infrastructure without having to modify the names of any of your resources.  If calling `Create` for a KMS Keyring fails for KeyRings that already exist on the server, you wouldn't be able to do this.

In either case, my intention was for this test to capture that logic, where I either expect a failure when re-creating, or for it to be handled gracefully.